### PR TITLE
Initial implementation of the incremental computation for map()

### PR DESCRIPTION
**Pull request by junghunyoo**
A new test SUITE for incremental computation testing was also added.

### CODE REVIEW

File: src/lasp_core.erl [hunk 1]
```diff
@@ -190,6 +191,32 @@ read(Id, Threshold, Store) ->
...
+            %% The first run uses the current value.
+            lasp:read(Id, undefined);
+        {strict, _} ->
+            %% From the second run, it always waits for a delta
+            %% (from the local or the remote).
+            Self = self(),
+            {ok, Object} = do(get, [Store, Id]),
+            WDT = lists:append(Object#dv.waiting_delta_threads, [{delta, Self}]),
+            do(put, [Store, Id, Object#dv{waiting_delta_threads=WDT}]),
+            receive
```
Comment (cmeiklejohn) [inline src/lasp_core.erl]: Eventually, this should be parameterized as the other ```read``` operation is.  Maybe leave a ```@todo``` here so we remember?

File: src/lasp_core.erl [hunk 2]
```diff
@@ -701,7 +732,14 @@ map(Id, Function, AccId, Store, BindFun, ReadFun) ->
                                  lasp_orset ->
                                      lasp_orset:map(Function, V)
                              end,
-                  {ok, _} = BindFun(AccId, AccValue, Store)
+                  {ok, _} = BindFun(AccId, AccValue, Store);
+             %% A delta from the input will generate a new delta for the output
```
Comment (cmeiklejohn) [inline src/lasp_core.erl]: Can you clarify what this comment means?

**Comment by cmeiklejohn** [comment]
Do you have any way to gauge the performance difference here?

File: src/lasp_plumtree_broadcast_distribution_backend.erl [hunk 3]
```diff
@@ -387,7 +391,12 @@ intersection(Left, Right, Intersection) ->
 %%
 -spec map(id(), function(), id()) -> ok | error().
 map(Id, Function, AccId) ->
-    gen_server:call(?MODULE, {map, Id, Function, AccId}, infinity).
+    case lasp_config:get(delta_mode, false) of
```
Comment (cmeiklejohn) [inline src/lasp_plumtree_broadcast_distribution_backend.erl]: Maybe we control this using another variable as well, like to enable or disable incremental computation?  We need a way to evaluate deltas with this enabled and with this disabled.

File: src/lasp_core.erl (hunk 2 above)
Comment (junghunyoo) [inline src/lasp_core.erl]: When the `process()` receives a delta of the input from  the `ReadFun()`, the `lasp_orset:map()` transforms it to a delta of the output. And `bind()` will add the delta to the output. This is what I want to say in that comment. I will update it :)

File: src/lasp_plumtree_broadcast_distribution_backend.erl (hunk 3 above)
Comment (junghunyoo) [inline src/lasp_plumtree_broadcast_distribution_backend.erl]: I will add a new variable for the incremental computation.

**Comment by junghunyoo** [comment]
In the `lasp_delta_SUITE`, there are two test cases. The first one is testing ORSet without the `delta_mode` and the incremental computation. And the second one enables the `delta_mode` and the incremental computation. Both cases measure the elapsed time for the `read()` of the output after making some changes to the input.

This is my latest results:
Time without incremental computation: 1262547
Time with incremental computation: 657246

One more thing... In order to see the benefits of the incremental computation, the original input set should be large enough (my test cases have 10,000 elements before making some changes to the input).

**Comment by cmeiklejohn** [comment]
That's total elapsed time, right?

It would be interesting to generate a graph of elapsed time for each operation -- each time an element is added the resulting transformation.  We *should* see based on that linear performance of the map O(n) and the improvement to O(1) when the incremental bits are enabled.

**Comment by junghunyoo** [comment]
The result time is only for the `read()` of the output after making an update for the input. I used the `timer:tc()` for it. You can check it in the 'lasp_delta_SUITE' file. Once I make the change for a new variable of the incremental computation, I will try to generate a graph with various tests.

**Comment by cmeiklejohn** [comment]
:+1:, good work!

**Pull request merged**